### PR TITLE
Add before change props

### PR DESCRIPTION
**Pull request by rijn**
#899

### CODE REVIEW

**Comment by lcx960324** [comment]
@rijn There are conflicts need to be dealt with. TAT

File: package.json [hunk 1]
```diff
@@ -71,7 +71,11 @@
     "gulp-less": "^3.3.2",
     "gulp-rename": "^1.2.2",
     "html-loader": "^0.3.0",
+    "html-webpack-plugin": "^2.28.0",
+    "is-promise": "^2.1.0",
+    "karma": "^1.4.1",
     "html-webpack-plugin": "^2.29.0",
+    "is-promise": "^2.1.0",
```
Comment (SergioCrisostomo) [inline package.json]: Duplicate `"is-promise": "^2.1.0",`

**Comment by SergioCrisostomo** [comment]
@icarusion  are you ok with this pull request?

File: src/components/switch/switch.vue [hunk 2]
```diff
@@ -56,11 +60,17 @@
                     return false;
                 }
 
-                const checked = !this.currentValue;
-                this.currentValue = checked;
-                this.$emit('input', checked);
-                this.$emit('on-change', checked);
-                this.dispatch('FormItem', 'on-form-change', checked);
+                checkPromise(this.beforeChange(this.currentValue)).then((result) => {
```
Comment (SergioCrisostomo) [inline src/components/switch/switch.vue]: Not sure I follow the logic here. If `beforeChange` is `null` then we get a error right?
Maybe should be 

    checkPromise(this.beforeChange)(this.currentValue).then((result) => {

?

Comment (rijn) [reply src/components/switch/switch.vue]: My bad. I didn't fully understand how promise works at that point. I'll fix it soon.

**Comment by icarusion** [comment]
I think there are many components that need to add `before-xxx` events, such as `Modal`、`Switch`、`Tabs`(when close), so i plan to make a list an discuss a unified promise solution.

File: src/components/switch/switch.vue (hunk 2 above)
Comment (SergioCrisostomo) [reply src/components/switch/switch.vue]: @rijn cool no problem. Maybe we will do this in more components later.

**Comment by SergioCrisostomo** [comment]
@icarusion good idea. I think also this is a good feature and sound good with a small discussion to find a unified promise solution. 

I'll close this for now then.

**Pull request closed without merging**